### PR TITLE
fix: search

### PR DESCRIPTION
**Pull request by ondrej-hosak**

### CODE REVIEW

File: lib/final-api/model_extensions.rb [hunk 1]
```diff
@@ -3,86 +3,138 @@
...
-    if (column == 'owner_id') || (column == 'stopped_by_id')
-      if operator == ':'
-        term = User.where("name ILIKE ?", "%#{term}%").pluck(:id)
+    # Returns list of parsed subqueries
+    #
+    # For example:
+    #   parse_query('nam:"foo bar baz" bui =qux id : 1')
+    #     => [ ['nam', ':', 'foo bar baz'], ['bui', '=', 'qux'], ['id', ':', '1']]
+    def parse_query(query)
+      expression_pattern = query_pattern
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb:54]: is this really needed?

File: lib/final-api/model_extensions.rb [hunk 2]
```diff
@@ -3,86 +3,138 @@
...
+      query = query.dup.strip
+
+      first_expression = query.slice(expression_pattern)
+
+      if first_expression.nil?
+        query = query.dup.strip
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb:60]: What's the meaning of this? Same code is few lines above.

File: lib/final-api/model_extensions.rb [hunk 3]
```diff
@@ -3,86 +3,138 @@
...
+        if query.starts_with?('"') && query.ends_with?('"')
+          query = query.dup.slice(1..-1)
+          query = query.dup.slice(0..-2)
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb:63]: this is hard to read

File: lib/final-api/model_extensions.rb [hunk 4]
```diff
@@ -3,86 +3,138 @@
...
-        term = User.where(name: term).pluck(:id)
+        check_query(query)
       end
-      operator = 'IN'
+
+      array = query.scan(expression_pattern)
+      return [['name', ':', query]] if array.empty?
+      wrong_keys = []
+      result = array.map do |item|
+        query_key = item[0].downcase
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb:73]: query_key, operator, query_value = item.flatten
query_key.downcase!

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+      end
+
+      context 'when id does not exist' do
+        it 'returns an empty array' do
+          expect(Build.search('id = \"99999999999999\" ', 5, 0).length).to eq(0)
+        end
+      end
+
+      context 'filters by StartedBy' do
+        it 'passes test 1' do
```
Comment (ondrej-hosak) [inline spec/features/lib/final-api/model_extensions_spec.rb:52]: passes test for user1

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+          builds << FactoryGirl.create(:build_customized, owner: user2)
+          filtered = Build.search("sta = #{user2.name}", 5, 0)
+          expect(filtered.first.owner.name).to eq(user2.name)
+          expect(filtered.length).to eq(1)
+        end
+
+        it 'passes test 2' do
```
Comment (ondrej-hosak) [inline spec/features/lib/final-api/model_extensions_spec.rb:59]: passes test for user2

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+          builds << FactoryGirl.create(:build_customized, owner: user2)
+          filtered = Build.search("sta = #{user1.name}", 500, 0)
+          expect(filtered.length).to eq(20)
+        end
+      end
+
+      it 'filters by stoppedBy' do
+        builds << FactoryGirl.create(:build_customized, stopped_by: user1, owner: user2)
+        filtered = Build.search('sto= "franta.lopata" ', 5, 0)
```
Comment (ondrej-hosak) [inline spec/features/lib/final-api/model_extensions_spec.rb:68]: user1.name

Comment (ireznice) [inline spec/features/lib/final-api/model_extensions_spec.rb:68]: DO NOT MERGE THIS!

**Pull request closed without merging**